atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -2374,6 +2353,22 @@ private boolean rollbackOtherTransaction(long startTs, TransactionService transa
         }
     }
 
+    ///////////////////////////////////////////////////////////////////////////
+    /// Close
+    ///////////////////////////////////////////////////////////////////////////
+
+    private void close() {
+        try {
+            // Run close() to release locks before running success callbacks, since success callbacks might
+            // start a new transaction and attempt to grab the same locks as the current transaction.
+            closer.close();
+        } finally {
+            if (isDefinitivelyCommitted()) {
+                successCallbackManager.runCallbacks();
+            }
```

Comment:
I think I was kinda alluding to this as well, running close once is a useful thing in general.
It almost feels like we'd want to centralize this behavior somewhere, we now have 2 different callbacks that we're running, could one be implemented on top of the other?

Reply:
AND it would be really cool to also centralize the feature that locks up pretty much all of the methods on Transaction (and makes them throw) AS SOON AS you start commit or abort. Right now it's very heavy handed and a lot of code to make sure you don't forget that, we could probably easily do that as a proxy or ForwardingTransaction